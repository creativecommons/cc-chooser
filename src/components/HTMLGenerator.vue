<template>
    <div>
        <!--
        <b-collapse aria-id='contentIdForA11y1'>
            <button
                class='button is-primary is-fullwidth collapsible-button'
                slot='trigger'
                aria-controls='contentIdForA11y1'>
                <h3 class='title is-3'></h3>
            </button>
                <div class='content'>-->
                    <h3 class='title is-3'>Embed your License Attribution!</h3>
                    <span id='attribution-richtext-container'>
                        <span id='attribution-richtext' class='photo_usage-attribution' ref='photoAttribution'>
                            <a :href='workLocation'
                                v-if='workTitle && workLocation'
                                target='_blank'
                                rel='noopener'>'{{ workTitle }}'
                            </a>
                            <a v-if='!workTitle && workLocation'
                                :href='workLocation'>
                                This work
                            </a>
                            <p v-if='!workTitle && !workLocation'>This work</p>
                            <p v-if='workTitle && !workLocation'>{{ workTitle }}</p>
                            <span v-if='attributeToName'>
                                by
                                <a v-if='attributeToURL'
                                    :href='attributeToURL'
                                    target='_blank'
                                    rel='noopener'>{{ attributeToName }}</a>
                                <span v-else>{{ attributeToName }}</span>
                            </span>
                            is licensed under
                            <a class='photo_license' :href='licenseURL' target='_blank' rel='noopener'>
                            {{ shortLicenseName }}
                            </a>
                        </span>
                        <LicenseIcons
                            :url='licenseURL'
                            :iconsArr='iconsArr'/>
                    </span>
                    <CopyButton id='copy-richtext-btn'
                                el='#attribution-richtext'
                                title='Copy the attribution to paste into your blog or document'>
                        Copy Rich Text
                    </CopyButton>
                    <div class='control' id='generated-html-container'>
                        <textarea id='attribution-html'
                            class='textarea'
                            :value='attributionHTML()'
                            readonly>
                        </textarea>
                    </div>
                    <CopyButton id='copy-html-btn'
                                el='#attribution-html'
                                title='Copy the attribution to paste into your blog or document'>
                        Copy HTML
                    </CopyButton>
                    <div id='generator-meta-inputs'>
                        <b id='generator-meta-inputs-heading'>
                            Filling out the boxes below is optional, but it adds more data
                            to the generated HTML and rich text above!
                        </b>
                        <div class='metadata-input-row'>
                            <div class='columns'>
                                <div class='column'>
                                    <b-field label='Title of Work'>
                                        <b-input
                                            v-model='workTitle'
                                            placeholder='workTitle'>
                                        </b-input>
                                    </b-field>
                                    <p class='help'>The title of the work being licensed.</p>
                                </div>
                                <div class='column'>
                                    <b-field label='Work Location'>
                                        <b-input
                                            v-model='workLocation'
                                            placeholder='workLocation'>
                                        </b-input>
                                    </b-field>
                                    <p class='help'>The URL where the work will be hosted or located.</p>
                                </div>
                            </div>
                        </div>
                        <div class='metadata-input-row'>
                            <div class='columns'>
                                <div class='column'>
                                    <b-field label='Attribute Work to URL'>
                                        <b-input
                                            v-model='attributeToURL'
                                            placeholder='attributeToURL'
                                        ></b-input>
                                    </b-field>
                                    <p class='help'>The URL to which the work should be attributed.</p>
                                </div>
                                <div class='column'>
                                    <b-field label='Attribute Work to Name'>
                                        <b-input
                                            v-model='attributeToName'
                                            placeholder='attributeToName'>
                                        </b-input>
                                    </b-field>
                                    <p class='help'>The name of the person who should receive attribution for the work.</p>
                                </div>
                            </div>
                        </div>
                    </div>
                <!--</div>
        </b-collapse>-->
    </div>
</template>
<script>
import CopyButton from './CopyButton'
import LicenseIcons from './LicenseIcons'
import LicenseIconography from '@creativecommons/vocabulary/vocabulary.common'
import attributionHtml from '@/utils/attributionHtml'
import licenseUrl from '@/utils/licenseUrl'

export default {
    name: 'HTMLGenerator',
    props: ['shortLicenseName'],
    components: {
        CopyButton,
        LicenseIcons,
        LicenseIconography
    },
    data() {
        return {
            workTitle: '',
            attributeToName: '',
            attributeToURL: '',
            workLocation: ''
        }
    },
    methods: {
        attributionHTML() {
            return attributionHtml({
                workLocation: this.workLocation,
                workTitle: this.workTitle,
                attributeToName: this.attributeToName,
                attributeToURL: this.attributeToURL
            },
            this.licenseURL,
            this.$props.shortLicenseName
            )
        }
    },
    computed: {
        licenseURL() {
            return licenseUrl(this.$props.shortLicenseName)
        },
        iconsArr() {
            return this.$props.shortLicenseName.toLowerCase().slice(3, this.$props.shortLicenseName.length - 4).split('-')
        }
    }
}
</script>
<style scoped>
    .collapsible-button {
        height: 70px;
        background-color: #ef9421 !important;
    }

    #generator-meta-inputs-heading {
        margin-bottom: 1rem;
        display: block;
    }

    b {
        text-align: center;
    }

    button {
        margin-top: .5rem !important
    }

    .field {
        margin-bottom: 0px;
    }

    #attribution-richtext-container {
        margin-top: 1rem;
        display: block;
    }

    #attribution-richtext-container .photo-license-icons {
        vertical-align: middle;
        margin-top: 0px;
        margin-left: 2px;
    }

    #attribution-richtext p {
        display: inline
    }

    #generated-html-container {
        margin-top: 1.5rem
    }

    textarea {
        word-break: break-all;
        -ms-word-break: break-all;
    }
</style>
