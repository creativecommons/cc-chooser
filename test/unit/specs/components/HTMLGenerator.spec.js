import HTMLGenerator from '@/components/HTMLGenerator'
import { shallowMount, createLocalVue } from '@vue/test-utils'
import Buefy from 'buefy'

const defaultHtml = '<p style="font-size: 0.9rem;font-style: italic;"><a href="">"This work"</a>' +
    ' is licensed under <a href="https://creativecommons.org/licenses/by/4.0/?ref=ccchooser" ' +
    'style="margin-right: 5px;">CC BY 4.0</a><a href="https://creativecommons.org/licenses/by/4.0/?ref=ccchooser" ' +
    'target="_blank" rel="noopener noreferrer" style="display: inline-block;white-space: none;opacity: .7;' +
    'margin-top: 2px;margin-left: 3px;height: 22px !important;"><img style="height: inherit;margin-right: 3px;' +
    'display: inline-block;" src="https://search.creativecommons.org/static/img/cc_icon.svg" />' +
    '<img style="height: inherit;margin-right: 3px;display: inline-block;" ' +
    'src="https://search.creativecommons.org/static/img/cc-by_icon.svg" /></a></p>'
const shortLicenseName = 'CC BY 4.0'
const defaultLicenseURL = 'https://creativecommons.org/licenses/by/4.0/?ref=ccchooser'

const newObject = {
    workTitle: 'newWorkTitle',
    attributeToName: 'newAttributeToName',
    attributeToURL: 'https://new_attribute_to_url.com',
    workLocation: 'https://new_work_location.com'
}

const localVue = createLocalVue()
localVue.use(Buefy)

describe('HTMLGenerator.vue', () => {
    const wrapper = shallowMount(HTMLGenerator, {
        localVue,
        propsData: {
            shortLicenseName
        }
    })

    it('renders the correct markup', () => {
        expect(wrapper.element).toMatchSnapshot()
        expect(wrapper.html()).toContain('<textarea id="attribution-html" readonly="readonly" class="textarea">')
        expect(wrapper.html().includes('textarea'))
    })

    it('renders correctly with default values', () => {
        expect(wrapper.vm.attributionHTML()).toEqual(defaultHtml)
        expect(wrapper.html().includes(shortLicenseName))
        expect(wrapper.vm.licenseURL).toEqual(defaultLicenseURL)
        expect(wrapper.vm.workTitle).toEqual('')
        expect(wrapper.vm.workLocation).toEqual('')
        expect(wrapper.vm.iconsArr).toEqual(['by'])
    })

    it('changes rich text correctly when workLocation changes', () => {
        wrapper.setData({'workLocation': newObject.workLocation})
        expect(wrapper.find('#attribution-richtext').html()).toContain(newObject.workLocation)
    })

    it('changes html correctly when title changes', () => {
        wrapper.setData({'workTitle': newObject.workTitle})
        expect(wrapper.vm.workTitle).toEqual(newObject.workTitle)
        expect(wrapper.vm.attributionHTML()).toContain(newObject.workTitle)
    })

    it('changes html correctly when attribute url and name fields change', () => {
        const byName = `<span> by <span>${newObject.attributeToName}</span></span>`
        const byNameAndURL = `<span> by <a href="${newObject.attributeToURL}">${newObject.attributeToName}</a></span>`
        wrapper.setData({'attributeToName': newObject.attributeToName})
        expect(wrapper.vm.attributionHTML()).toContain(byName)
        wrapper.setData({'attributeToURL': newObject.attributeToURL})
        expect(wrapper.vm.attributionHTML()).toContain(byNameAndURL)
    })

    it('changes correctly when license type changes', () => {
        const ncnd = 'CC BY-NC-ND 4.0'
        const ncndURL = 'https://creativecommons.org/licenses/by-nc-nd/4.0/?ref=ccchooser'
        const ncndIconsArr = ['by', 'nc', 'nd']
        wrapper.setProps({'shortLicenseName': ncnd})
        expect(wrapper.vm.attributionHTML()).toContain(ncndURL)
        expect(wrapper.html()).toContain(ncnd)
        expect(wrapper.vm.licenseURL).toEqual(ncndURL)
        expect(wrapper.vm.iconsArr).toEqual(ncndIconsArr)
    })
})
