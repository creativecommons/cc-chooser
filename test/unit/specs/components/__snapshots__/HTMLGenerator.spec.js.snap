// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HTMLGenerator.vue renders the correct markup 1`] = `
<div>
  <h3
    class="title is-3"
  >
    Embed your License Attribution!
  </h3>
   
  <span
    id="attribution-richtext-container"
  >
    <span
      class="photo_usage-attribution"
      id="attribution-richtext"
    >
      <!---->
       
      <!---->
       
      <p>
        This work
      </p>
       
      <!---->
       
      <!---->
      
                        is licensed under
                        
      <a
        class="photo_license"
        href="https://creativecommons.org/licenses/by/4.0/?ref=ccchooser"
        rel="noopener"
        target="_blank"
      >
        
                        CC BY 4.0
                        
      </a>
    </span>
     
    <licenseicons-stub
      iconsarr="by"
      url="https://creativecommons.org/licenses/by/4.0/?ref=ccchooser"
    />
  </span>
   
  <copybutton-stub
    el="#attribution-richtext"
    id="copy-richtext-btn"
    title="Copy the attribution to paste into your blog or document"
  >
    
                    Copy Rich Text
                
  </copybutton-stub>
   
  <div
    class="control"
    id="generated-html-container"
  >
    <textarea
      class="textarea"
      id="attribution-html"
      readonly="readonly"
    />
  </div>
   
  <copybutton-stub
    el="#attribution-html"
    id="copy-html-btn"
    title="Copy the attribution to paste into your blog or document"
  >
    
                    Copy HTML
                
  </copybutton-stub>
   
  <div
    id="generator-meta-inputs"
  >
    <b
      id="generator-meta-inputs-heading"
    >
      
                        Filling out the boxes below is optional, but it adds more data
                        to the generated HTML and rich text above!
                    
    </b>
     
    <div
      class="metadata-input-row"
    >
      <div
        class="columns"
      >
        <div
          class="column"
        >
          <b-field-stub
            addons="true"
            label="Title of Work"
          >
            <b-input-stub
              customclass=""
              hascounter="true"
              placeholder="workTitle"
              type="text"
              usehtml5validation="true"
              value=""
            />
          </b-field-stub>
           
          <p
            class="help"
          >
            The title of the work being licensed.
          </p>
        </div>
         
        <div
          class="column"
        >
          <b-field-stub
            addons="true"
            label="Work Location"
          >
            <b-input-stub
              customclass=""
              hascounter="true"
              placeholder="workLocation"
              type="text"
              usehtml5validation="true"
              value=""
            />
          </b-field-stub>
           
          <p
            class="help"
          >
            The URL where the work will be hosted or located.
          </p>
        </div>
      </div>
    </div>
     
    <div
      class="metadata-input-row"
    >
      <div
        class="columns"
      >
        <div
          class="column"
        >
          <b-field-stub
            addons="true"
            label="Attribute Work to URL"
          >
            <b-input-stub
              customclass=""
              hascounter="true"
              placeholder="attributeToURL"
              type="text"
              usehtml5validation="true"
              value=""
            />
          </b-field-stub>
           
          <p
            class="help"
          >
            The URL to which the work should be attributed.
          </p>
        </div>
         
        <div
          class="column"
        >
          <b-field-stub
            addons="true"
            label="Attribute Work to Name"
          >
            <b-input-stub
              customclass=""
              hascounter="true"
              placeholder="attributeToName"
              type="text"
              usehtml5validation="true"
              value=""
            />
          </b-field-stub>
           
          <p
            class="help"
          >
            The name of the person who should receive attribution for the work.
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
